rpcs3/Emu/RSX/GL/GLHelpers.cpp

```diff
@@ -52,33 +52,119 @@ namespace gl
+			break;
+		// Generated by the user of this application
+		case GL_DEBUG_SOURCE_APPLICATION:
+			str_source = "APPLICATION";
+			break;
+		// Some source that isn't one of these
+		case GL_DEBUG_SOURCE_OTHER:
+			str_source = "OTHER";
+			break;
+		// Not on documentation
+		default:
+			str_source = "UNKNOWN";
+			rsx_log.error("log_debug(source=%d): Unknown message source", source);
+		}
+
+		// Message type
+		std::string str_type;
 		switch (type)
 		{
+		// An error, typically from the API
 		case GL_DEBUG_TYPE_ERROR:
-		{
-			rsx_log.error("%s", message);
-			return;
+			str_type = "ERROR";
+			break;
+		// Some behavior marked deprecated has been used
+		case GL_DEBUG_TYPE_DEPRECATED_BEHAVIOR:
+			str_type = "DEPRECATED";
+			break;
+		// Something has invoked undefined behavior
+		case GL_DEBUG_TYPE_UNDEFINED_BEHAVIOR:
+			str_type = "UNDEFINED";
+			break;
+		// Some functionality the user relies upon is not portable
+		case GL_DEBUG_TYPE_PORTABILITY:
+			str_type = "PORTABILITY";
+			break;
+		// Code has triggered possible performance issues
+		case GL_DEBUG_TYPE_PERFORMANCE:
+			str_type = "PERFORMANCE";
+			break;
+		// Command stream annotation
+		case GL_DEBUG_TYPE_MARKER:
+			str_type = "MARKER";
+			break;
+		// Group pushing
+		case GL_DEBUG_TYPE_PUSH_GROUP:
+			str_type = "PUSH_GROUP";
+			break;
+		// foo
+		case GL_DEBUG_TYPE_POP_GROUP:
+			str_type = "POP_GROUP";
+			break;
+		// Some type that isn't one of these
+		case GL_DEBUG_TYPE_OTHER:
+			str_type = "OTHER";
+			break;
+		// Not on documentation
+		default:
+			str_type = "UNKNOWN";
+			rsx_log.error("log_debug(type=%d): Unknown message type", type);
 		}
+
+	  switch (severity)
+	  {
+		// All OpenGL Errors, shader compilation/linking errors, or highly-dangerous undefined behavior
+		case GL_DEBUG_SEVERITY_HIGH:
+		// Major performance warnings, shader compilation/linking warnings, or the use of deprecated functionality
+		case GL_DEBUG_SEVERITY_MEDIUM:
+			rsx_log.error("[DEBUG_OUTPUT] [%s] [%s] [%d]: %s", str_source, str_type, id, message);
+			return;
+		// Redundant state change performance warning, or unimportant undefined behavior
+		case GL_DEBUG_SEVERITY_LOW:
+			rsx_log.warning("[DEBUG_OUTPUT] [%s] [%s] [%d]: %s", str_source, str_type, id, message);
+			return;
+		// Anything that isn't an error or performance issue
+		case GL_DEBUG_SEVERITY_NOTIFICATION:
+			rsx_log.notice("[DEBUG_OUTPUT] [%s] [%s] [%d]: %s", str_source, str_type, id, message);
+			return;
+		// Not on documentation
 		default:
-		{
-			rsx_log.warning("%s", message);
+			rsx_log.error("log_debug(severity=%d): Unknown severity level", severity);
+			rsx_log.error("[DEBUG_OUTPUT] [%s] [%s] [%d]: %s", str_source, str_type, id, message);
 			return;
-		}
-		}
+	  }
 	}
-#endif
 
 	void enable_debugging()
 	{
-#ifdef WIN32
-		glEnable(GL_DEBUG_OUTPUT_SYNCHRONOUS);
-		glDebugMessageCallback(static_cast<GLDEBUGPROC>(dbgFunc), nullptr);
-#endif
+	  glEnable(GL_DEBUG_OUTPUT);
```

Comment:
Has to be synchronous as we use breakpoints to actually see where the error is coming from on our side.